src/Database/Esqueleto/Record.hs

```diff
@@ -379,22 +383,81 @@ sqlSelectProcessRowDec RecordInfo {..} = do
-sqlSelectProcessRowPat fieldType var =
-  case fieldType of
-    -- Entity x -> var
-    AppT (ConT ((==) ''Entity -> True)) _innerType -> VarP var
-    -- Maybe (Entity x) -> var
-    AppT
-      (ConT ((==) ''Maybe -> True))
-      (AppT (ConT ((==) ''Entity -> True)) _innerType) -> VarP var
-    -- x -> Value var
+-- * If there exists an instance @'SqlSelect' sql x@, then a type of @x@ gives a pattern of @var@.
+--
+-- This function should match `sqlFieldType`.
+sqlSelectProcessRowPat :: Type -> Name -> Q Pat
+sqlSelectProcessRowPat fieldType var = do
+  maybeSqlType <- reifySqlSelectType fieldType
+
+  case maybeSqlType of
+    Just _ -> pure $ VarP var
+    Nothing -> case fieldType of
+        -- Entity x -> var
+        AppT (ConT ((==) ''Entity -> True)) _innerType -> pure $ VarP var
+        -- Maybe (Entity x) -> var
+        (ConT ((==) ''Maybe -> True))
+          `AppT` ((ConT ((==) ''Entity -> True))
+                  `AppT` _innerType) -> pure $ VarP var
+        -- x -> Value var
 #if MIN_VERSION_template_haskell(2,18,0)
-    _ -> ConP 'Value [] [VarP var]
+        _ -> pure $ ConP 'Value [] [VarP var]
 #else
-    _ -> ConP 'Value [VarP var]
+        _ -> pure $ ConP 'Value [VarP var]
 #endif
 
+-- Given a type, find the corresponding SQL type.
+--
+-- If there exists an instance `SqlSelect sql ty`, then the SQL type for `ty`
+-- is `sql`.
+--
+-- This function definitely works for records and instances generated by this
+-- module, and might work for instances outside of it.
+reifySqlSelectType :: Type -> Q (Maybe Type)
+reifySqlSelectType originalType = do
+  -- Here we query the compiler for Instances of `SqlSelect a $(originalType)`;
+  -- the API for this is super weird, it interprets a list of types as being
+  -- applied as successive arguments to the typeclass name.
+  --
+  -- See: https://gitlab.haskell.org/ghc/ghc/-/issues/21825
+  --
+  -- >>> reifyInstances ''SqlSelect [VarT (mkName "a"), ConT ''MyRecord]
+  -- [ InstanceD Nothing
+  --             []
+  --             (AppT (AppT (ConT Database.Esqueleto.Internal.Internal.SqlSelect)
+  --                         (ConT Ghci3.SqlMyRecord))
+  --                   (ConT Ghci3.MyRecord))
+  --             []
+  -- ]
+  tyVarName <- newName "a"
+  instances <- reifyInstances ''SqlSelect [VarT tyVarName, originalType]
+
+  -- Given the original type (`originalType`) and an instance type for a
+  -- `SqlSelect` instance, get the SQL type which corresponds to the original
+  -- type.
+  let extractSqlRecord :: Type -> Type -> Maybe Type
+      extractSqlRecord originalTy instanceTy =
+        case instanceTy of
+          (ConT ((==) ''SqlSelect -> True))
+            `AppT` sqlTy
+            `AppT` ((==) originalTy -> True) -> Just sqlTy
+          _ -> Nothing
+
+      -- Filter `instances` to the instances which match `originalType`.
+      filteredInstances :: [Type]
+      filteredInstances =
+        flip mapMaybe instances
```

Comment:
hm, `listToMaybe . mapMaybe f` is also `find (isJust . f)` - may be easier or more concise to write like that?

Reply:
I don't think so, because `find` only returns the element but I also need the `mapMaybe` part to transform and extract the sql type.

Reply:
Oh, good point!